pkg/controller/session/session_controller.go

```diff
@@ -70,80 +91,168 @@ func (r *ReconcileSession) Reconcile(request reconcile.Request) (reconcile.Resul
+	refs := statusesToRef(*session)
+	if deleted {
+		deleteAllRefs(ctx, session, refs)
+	} else {
+		if err := syncAllRefs(ctx, session); err != nil {
+			return reconcile.Result{}, err
+		}
+		deleteRemovedRefs(ctx, session, refs)
+	}
 
-	deployment := appsv1.Deployment{}
+	if deleted {
+		session.RemoveFinalizer(finalizer)
+		if err := r.client.Update(ctx, session); err != nil {
+			ctx.Log.Error(err, "Failed to remove finalizer on session")
+		}
 
-	err = r.client.Get(ctx, types.NamespacedName{Namespace: request.Namespace, Name: instance.Spec.Ref}, &deployment)
-	if err != nil {
-		updateStatus(ctx, reqLogger, r.client, instance, fmt.Sprintf("%v", err))
-		return reconcile.Result{Requeue: false}, err
 	}
+	return reconcile.Result{}, nil
+}
 
-	reqLogger.Info("Found Deployment", "image", deployment.Spec.Template.Spec.Containers[0].Image)
-
-	serviceList := corev1.ServiceList{}
-	err = r.client.List(ctx, &client.ListOptions{Namespace: request.Namespace}, &serviceList)
-	if err != nil {
-		updateStatus(ctx, reqLogger, r.client, instance, fmt.Sprintf("%v", err))
-		return reconcile.Result{Requeue: false}, err
+func deleteRemovedRefs(ctx model.SessionContext, session *istiov1alpha1.Session, refs []*model.Ref) { //nolint[:hugeParam]
+	for _, ref := range refs {
+		found := false
+		for _, r := range session.Spec.Refs {
+			if ref.Name == r {
+				found = true
+			}
+		}
+		if !found {
+			if err := delete(ctx, session, ref); err != nil {
+				ctx.Log.Error(err, "Failed to delete session ref", "ref", ref)
+			}
+		}
 	}
-	for i := range serviceList.Items {
-		vs := &serviceList.Items[i]
-		reqLogger.Info("Found Service", "name", vs.ObjectMeta.Name, "labels", vs.Labels)
+}
+func deleteAllRefs(ctx model.SessionContext, session *istiov1alpha1.Session, refs []*model.Ref) { //nolint[:hugeParam]
+	for _, ref := range refs {
+		if err := delete(ctx, session, ref); err != nil {
+			ctx.Log.Error(err, "Failed to delete session ref", "ref", ref)
+		}
 	}
+}
 
-	drList := istionetwork.DestinationRuleList{}
-	err = r.client.List(ctx, &client.ListOptions{Namespace: request.Namespace}, &drList)
-	if err != nil {
-		updateStatus(ctx, reqLogger, r.client, instance, fmt.Sprintf("%v", err))
-		return reconcile.Result{Requeue: false}, err
+func delete(ctx model.SessionContext, session *istiov1alpha1.Session, ref *model.Ref) error { //nolint[:hugeParam]
+	ctx.Log.Info("Remove ref", "name", ref.Name)
+
+	statusToRef(*session, ref)
+	for _, revertor := range revertors {
+		err := revertor(ctx, ref)
+		if err != nil {
+			ctx.Log.Error(err, "Revert", "name", ref.Name)
+		}
 	}
-	for i := range drList.Items {
-		vs := &drList.Items[i]
-		reqLogger.Info("Found DestinationRule", "name", vs.ObjectMeta.Name)
+	refToStatus(*ref, session)
+	return ctx.Client.Status().Update(ctx, session)
+}
+
+func syncAllRefs(ctx model.SessionContext, session *istiov1alpha1.Session) error { //nolint[:hugeParam]
```

Comment:
We seem to have bunch of `Refs` related logic in here (as well as ref <-> status conversion) - would it make sense to extract it to its own file(s) for example?

Reply:
On top of that what are your thoughts about unit testing this?

Reply:
https://github.com/aslakknutsen/istio-workspace/issues/72